### PR TITLE
Added Google Login functionality and tables.

### PR DESCRIPTION
**Pull request by KennethLundberg**

### CODE REVIEW

File: ucsdscheduleme/Controllers/AccountController.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        [HttpGet]
+        [AllowAnonymous]
+        public async Task<IActionResult> ExternalLoginCallback(string returnUrl = null, string remoteError = null)
+        {
+            if (remoteError != null)
+            {
+                ErrorMessage = $"Error from external provider: {remoteError}";
+                return RedirectToAction(nameof(Login));
+            }
```
Comment (harishduwadi) [inline ucsdscheduleme/Controllers/AccountController.cs]: remoteError = null in parameter, but we're checking if remoteError is null or not. It's a little confusing why we're checking it.

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+        public static ScheduleViewModel FormatSectionsToCalendarEvent(List<Section> sections)
+        {
+
+            ScheduleViewModel model = new ScheduleViewModel();
+            List<CalendarEvent> events = new List<CalendarEvent>();
+            List<OneTimeEvent> oneTimeEvents = new List<OneTimeEvent>();
+            List<Metadata> classMetadata = new List<Metadata>();
+            Metadata overall = new Metadata();
+
```
Comment (cjimene) [inline ucsdscheduleme/Repo/FormatRepo.cs]: Add comments to the effect of "initializing locals" and "defining objects" or something like that, just for ease of skimming.

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using ucsdscheduleme.Models;
+
+namespace ucsdscheduleme.Repo
+{
+    public class FormatRepo
+    {
+        private static readonly int START_HOUR = 7;
+        private static readonly int START_MINUTES = 30;
```
Comment (harishduwadi) [inline ucsdscheduleme/Repo/FormatRepo.cs]: It's a little ambiguous why we've 7 and 30 of these constants

File: ucsdscheduleme/Models/Meeting.cs [hunk 4]
```diff
@@ -28,8 +28,8 @@ public class Meeting
         [StringLength(3)]
         public string Code { get; set; }
         public Days Days { get; set; }
-        public DateTime? StartTime { get; set; }
-        public DateTime? EndTime { get; set; }
+        public DateTime StartTime { get; set; }
```
[the remark this answers is not in the document]
Comment (KennethLundberg) [reply ucsdscheduleme/Models/Meeting.cs]: DateTime functions arn't available when you have a nullable DateTime, and the code to fix this at each location is a lot of boiler plate, its much easier to validate the date to be valid at the start of optimization, and in all other cases assume you have a start date and end date with correct values by the time you get to the point where you need them.

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        private static readonly int MINUTES_IN_HOUR = 60;
+
+        public static ScheduleViewModel FormatSectionsToCalendarEvent(List<Section> selectedSections)
```
Comment (git-ram) [inline ucsdscheduleme/Repo/FormatRepo.cs]: Comments on methods helps

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                CourseAbbreviation = courseAbbreviation,
+                DateAndTime = start,
+                Location = $"{meeting.Location.RoomNumber} {meeting.Location.Building}",
+                Type = meeting.MeetingType.ToString()
+            };
+
+            oneTimeEvent.Add(final);
+        }
+
+        private static void AddCalendarEvent(ref List<CalendarEvent> events, string courseAbbreviation, string professorName, Meeting meeting)
```
Comment (git-ram) [inline ucsdscheduleme/Repo/FormatRepo.cs]: Would be better if you could avoid long lines like 187

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        {
+            DateTime start = meeting.StartTime;
+            string sectionCode = meeting.Code;
+            MeetingType type = meeting.MeetingType;
+
+            // Formats it into the form: 12:00am
+            string startString = start.ToString("h:mmtt");
+            string endString = meeting.EndTime.ToString("h:mmtt");
+
+            int durationInMinutes = (int)meeting.EndTime.Subtract(start).TotalMinutes;
```
Comment (seannam) [inline ucsdscheduleme/Repo/FormatRepo.cs]: whats the benefit of using .Substract() over ' - '?

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+            Dictionary<int, List<CalendarEvent>> thisBasesSectionEvents = PopulateSectionEventsForBase(sectionsForBase, course);
+            thisBase.SectionEvents = thisBasesSectionEvents;
+
+            return thisBase;
+        }
+
+        private static Dictionary<int, List<CalendarEvent>> PopulateSectionEventsForBase(List<Section> sectionsForBase, Course course)
+        {
+            Dictionary<int, List<CalendarEvent>> thisBasesSectionEvents = new Dictionary<int, List<CalendarEvent>>();
```
Comment (seannam) [inline ucsdscheduleme/Repo/FormatRepo.cs]: explaining what the key value is would be helpful even if you can tell from the following code like line 80

File: ucsdscheduleme/Repo/FormatRepo.cs [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                    Bases = new Dictionary<char, BaseViewModel>(),
+                    SelectedSection = selectedSection.Id,
+                    // If our sections have no meetings, we're in trouble. Time to abort.
+                    SelectedBase = selectedSection.Meetings?.First().Code[0] ?? 
+                        throw new ArgumentException($"Section with no meeting found: Id'{selectedSection.Id}'")
+                };
+
+                model.Courses.Add(course.Id,thisCourse);
+
+                var baseSectionGroups = course.Sections.GroupBy(s => s.Meetings.First().Code[0]);
```
Comment (cjimene) [inline ucsdscheduleme/Repo/FormatRepo.cs]: Slight overuse of var where datatype declarations would be useful for cleanliness.

**Pull request merged**